src/renderer/components/+namespaces/namespace.store.ts

```diff
@@ -1,47 +1,46 @@
-import { action, observable, reaction } from "mobx";
+import { action, comparer, observable, reaction } from "mobx";
 import { autobind, createStorage } from "../../utils";
 import { KubeObjectStore } from "../../kube-object.store";
 import { Namespace, namespacesApi } from "../../api/endpoints";
-import { IQueryParams, navigation, setQueryParams } from "../../navigation";
+import { createUrlParam } from "../../navigation";
 import { apiManager } from "../../api/api-manager";
 import { isAllowedResource } from "../../../common/rbac";
 import { getHostedCluster } from "../../../common/cluster-store";
 
+const storage = createStorage<string[]>("context_namespaces", []);
+
+export const namespaceUrlParam = createUrlParam<string[]>({
+  name: "namespaces",
+  isSystem: true,
+  multiValues: true,
+  get defaultValue() {
+    return storage.get();
+  }
+});
+
 @autobind()
 export class NamespaceStore extends KubeObjectStore<Namespace> {
   api = namespacesApi;
-  contextNs = observable.array<string>();
-
-  protected storage = createStorage<string[]>("context_ns", this.contextNs);
-
-  get initNamespaces() {
-    const fromUrl = navigation.searchParams.getAsArray("namespaces");
-
-    return fromUrl.length ? fromUrl : this.storage.get();
-  }
+  contextNs = observable.array<string>(storage.get());
 
   constructor() {
     super();
+    this.init();
+  }
 
-    // restore context namespaces
-    const { initNamespaces: namespaces } = this;
-
-    this.setContext(namespaces);
-    this.updateUrl(namespaces);
+  private init() {
+    // setup initial context namespaces from URL (when provided) or local-storage (default)
+    this.setContext(namespaceUrlParam.get());
 
-    // sync with local-storage & url-search-params
-    reaction(() => this.contextNs.toJS(), namespaces => {
-      this.storage.set(namespaces);
-      this.updateUrl(namespaces);
+    return reaction(() => this.contextNs.toJS(), namespaces => {
+      storage.set(namespaces); // save to local-storage
+      namespaceUrlParam.set(namespaces, { replaceHistory: true }); // update url
+    }, {
+      fireImmediately: true,
+      equals: comparer.identity,
     });
   }
 
-  getContextParams(): Partial<IQueryParams> {
-    return {
-      namespaces: this.contextNs
-    };
-  }
```

Comment:
So this is a breaking change. We do export the `NamespaceStore`. This either has to be kept (since we are doing this release after 4.0.0) or this store has to be marked as a beta version.

Reply:
It's surprising to me that this part publishly exposed to extensions-api (which should not and if needed better to add some facade).